### PR TITLE
Optimize cut point selection

### PR DESCRIPTION
**Pull request by dodgejoel**
This should address #27.

The approach that I took was to be slightly more careful about which cut points should be considered. A cut point is skipped over if:
1. it does not correspond to a change in col-value, or 
2. the labels for the previous col-value and the labels for the current col-value are identical and consist of exactly one label.

This is a reasonable condition to use here because it is precisely this configuration of label values which ensures that the cut point in question either does *not* come from a choice of threshold value for cutting or is *not* a so-called boundary point (in the language of the original paper).

Making these changes to the logic in find_cut_points necessitated injecting some auxiliary data into the function.

In addition, I have refactored how the iris test result data to make them easier to update and reuse. Now they stored in a csv instead of inlined in python. I have also included some testing util functions for loading and updating these data.

### CODE REVIEW

File: mdlp/_mdlp.pyx [hunk 1]
```diff
@@ -129,9 +136,16 @@ def find_cut(np.ndarray[np.int64_t, ndim=1] y, int start, int end):
 
     Parameters
     ----------
+    col: 1-d numpy array
+        column values
+
     y : 1-d numpy array
         class labels
 
+    labels_map: dict
```
Comment (hlin117) [inline mdlp/_mdlp.pyx:145]: I highly recommend you do not use a python `dict` in Cython code. One of the biggest reasons we code in Cython is so we can avoid the python GIL on critical code paths. So by using a `dict`, you're actually defeating the purpose of entering Cython land.

See https://stackoverflow.com/a/32267243/2014591.

File: tests/test_mdlp.py [hunk 2]
```diff
@@ -23,13 +26,17 @@ def test_slice_entropy():
...
 
 def test_find_cut_no_cut():
+    x = np.array([0., 1., 2., 3.])
     y = np.array([0, 0, 0, 0])
-    k = find_cut(y, 0, len(y))
+    labels_map = {0: {0}, 1: {0}, 2: {0}, 3: {0}}
+    k = find_cut(x, y, labels_map, 0, len(y))
     assert_equal(-1, k)
 
 def test_mdlp_iris():
```
Comment (hlin117) [inline tests/test_mdlp.py:42]: To be honest, I never liked this test in the first place. When I wrote this code, I wasn't aware of good testing practices. If I were to rewrite this code again, I would use a much smaller dataset to test with (maybe a `4 x 2` array?)

File: mdlp/_mdlp.pyx [hunk 3]
```diff
@@ -113,7 +120,7 @@ cdef bint reject_split(np.ndarray[np.int64_t, ndim=1] y, int start, int end, int
     return gain <= 1 / N * (log(N - 1) + delta)
 
 @cython.boundscheck(False)
-def find_cut(np.ndarray[np.int64_t, ndim=1] y, int start, int end):
+def find_cut(np.ndarray[np.float64_t, ndim=1] col, np.ndarray[np.int64_t, ndim=1] y, dict labels_map, int start, int end):
     """Finds the best cut between the specified interval.
 
     If k split the array into two sub arrays A and B, then
```
Comment (hlin117) [inline mdlp/_mdlp.pyx:126]: I need to freshen up this documentation when I have the time. "Best cut" is too vague. The definition is really taken from the original paper. (MDLPC Criteron.)

**Review by hlin117 — COMMENTED** [review]
A couple of issues with this PR:

1. You shouldn't use a dict in Cython
2. Your test case still uses the Iris dataset, so you're finding the need to import test fixtures
3. You're introducing breaking changes for those who have the package already. My suggestion is to pass a parameter in indicating whether they would like to use this optimization criterion. 

I'm also convinced that there's an easier way to do this, but I need to think about this problem more closely.

**Review by hlin117 — CHANGES_REQUESTED** [review]
Can you break your PR into several parts? I would like to stop using the entire iris dataset as a unit test. Unit tests should be small, and easy to understand.

**Comment by dodgejoel** [comment]
In response to your comments:
1. sounds good I will rework this to avoid using a dict. Your reasoning makes sense.
2. Sounds good. Following your suggestion I will make a smaller test data.
3. I think that breaking changes are different from just changes that alter behavior. It is true that someone performing discretization after this change would get a different result than what they get before this change but does that mean that it is a 'breaking' change? Adding a parameter to control whether or not every improvement will be used seems like a recipe for eventually overwhelming technical debt in this project - should it live so long and be so used :)

"Can you break your PR into several parts? I would like to stop using the entire iris dataset as a unit test. Unit tests should be small, and easy to understand."

Happy to do this. I will submit PRs to 1. modify the test suite as you suggest and then 2. update find_cut_points logic similar to what I have submitted in this PR. 

A question about github workflows: Once this branch has been developed to this point and submitted as a PR - what is the right way to refactor into multiple PRs? Do I manage separate local branches for each PR that I will submit and then submit them one at a time?  I am used to gerrit workflows so please excuse my ignorance here. Thanks!

**Comment by dodgejoel** [comment]
Changes to test suite that you requested are in #29. I will plan on closing this pull request and pushing changes that you requested in separate pull requests that are curated a little more carefully to be modestly scoped and not bite off too many changes at once.

**Pull request closed without merging**